src/mito2/src/memtable/bulk/part.rs

```diff
@@ -0,0 +1,666 @@
+#[allow(clippy::too_many_arguments)]
+fn sort_arrays_to_record_batch(
+    metadata: &RegionMetadataRef,
+    pk: BinaryArray,
+    timestamp: ArrayRef,
+    sequence: UInt64Array,
+    op_type: UInt8Array,
+    fields: impl Iterator<Item = ArrayRef>,
+    schema: SchemaRef,
+    dedup: bool,
+) -> Result<(RecordBatch, i64, i64)> {
+    let timestamp_iter = timestamp_array_to_iter(metadata, &timestamp);
+    let (mut min_timestamp, mut max_timestamp) = (i64::MAX, i64::MIN);
+    let mut to_sort = pk
+        .iter()
+        .zip(timestamp_iter)
+        .zip(sequence.iter())
+        .map(|((pk, timestamp), sequence)| {
+            max_timestamp = max_timestamp.max(*timestamp);
+            min_timestamp = min_timestamp.min(*timestamp);
+            (pk, timestamp, sequence)
+        })
+        .enumerate()
+        .collect::<Vec<_>>();
+
+    to_sort.sort_unstable_by(|(_, (l_pk, l_ts, l_seq)), (_, (r_pk, r_ts, r_seq))| {
+        l_pk.cmp(r_pk)
+            .then(l_ts.cmp(r_ts))
+            .then(l_seq.cmp(r_seq).reverse())
+    });
+
+    if dedup {
+        // Dedup by timestamps while ignore sequence.
+        to_sort.dedup_by(|(_, (l_pk, l_ts, _)), (_, (r_pk, r_ts, _))| l_pk == r_pk && l_ts == r_ts);
+    }
+
+    let indices = UInt32Array::from_iter_values(to_sort.iter().map(|v| v.0 as u32));
+
+    let pk_dictionary = Arc::new(binary_array_to_dictionary(
+        // safety: pk must be BinaryArray
+        arrow::compute::take(
+            &pk,
+            &indices,
+            Some(TakeOptions {
+                check_bounds: false,
+            }),
+        )
+        .context(ComputeArrowSnafu)?
+        .as_any()
+        .downcast_ref::<BinaryArray>()
+        .unwrap(),
+    )?) as ArrayRef;
+    let mut arrays = Vec::with_capacity(schema.fields.len());
+    for arr in fields {
+        arrays.push(
+            arrow::compute::take(
+                &arr,
+                &indices,
+                Some(TakeOptions {
+                    check_bounds: false,
+                }),
+            )
+            .context(ComputeArrowSnafu)?,
+        );
+    }
+
+    let timestamp = arrow::compute::take(
+        &timestamp,
+        &indices,
+        Some(TakeOptions {
+            check_bounds: false,
+        }),
+    )
+    .context(ComputeArrowSnafu)?;
+
+    arrays.push(timestamp);
+    arrays.push(pk_dictionary);
+    arrays.push(
+        arrow::compute::take(
+            &sequence,
+            &indices,
+            Some(TakeOptions {
+                check_bounds: false,
+            }),
+        )
+        .context(ComputeArrowSnafu)?,
+    );
+
+    arrays.push(
+        arrow::compute::take(
+            &op_type,
+            &indices,
+            Some(TakeOptions {
+                check_bounds: false,
+            }),
+        )
+        .context(ComputeArrowSnafu)?,
+    );
+
+    let batch = RecordBatch::try_new(schema, arrays).context(NewRecordBatchSnafu)?;
```

Comment:
Is it possible that the batch is empty? We will also get incorrect min/max timestamps.

Reply:
Added length check here https://github.com/GreptimeTeam/greptimedb/pull/4163/commits/e46d05fb6a67dd865b8f1f02f15e46e82c3ed1cd#diff-176b5ab04e2d4dfe92ae84d90f94a03fe5e9eb2bf721511cd991d47a8408548cR223